### PR TITLE
Implement ind_from_latlon using a BallTree

### PR DESCRIPTION
**Pull request by leuty**
I tried using ind_from_latlon on with the new ICON-1 domain of MCH and it was kinda slow (20 Min). I implemented a new version using a BallTree.

I also added functionality to return the n closest points. That is super useful if one want to compare against instruments and avoid double penalties by slight location offsets.

### CODE REVIEW

**Comment by leuty** [comment]
Note, an version for spherical earth can be found here:

https://github.com/Unidata/python-workshop/blob/fall-2016/notebooks/netcdf-by-coordinates.ipynb

**Comment by AnnikaLau** [comment]
Thanks for the update/improvement! I'm still going through it but one thing I'm not sure of yet, would `ind[0]` be the closest of the closest points?

**Comment by leuty** [comment]
@AnnikaLau 
 I think the distinction is made by either passing `k=n`  or `k=[n]` to `cKDTree.query`, see [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html).

Would it help if I re-implemented the `verbose` option for `ind_from_latlon` ?

**Comment by AnnikaLau** [comment]
> @AnnikaLau I think the distinction is made by either passing `k=n` or `k=[n]` to `cKDTree.query`, see [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html).
> 
> Would it help if I re-implemented the `verbose` option for `ind_from_latlon` ? Then you can test.

Yes I think the `verbose` option is useful for the case k=1, which we use for the icon-vis examples. So that the users are aware of that the location of the nearest point might be quite different. So it would be great if you could re-implement it for that case.

**Comment by leuty** [comment]
@AnnikaLau 
`So that the users are aware of that the location of the nearest point might be quite different.`

I dont understand. Does it yield a different location for your case?

File: iconarray/core/utilities.py [hunk 1]
```diff
@@ -35,60 +38,72 @@ def awhere_drop(ds, cond):
...
     if verbose:
-        print(f"Closest ind: {ind}")
-        print(f" Given lat: {lat:.3f} vs found lat: {lats[ind]:.3f}")
-        print(f" Given lon: {lon:.3f} vs found lon: {lons[ind]:.3f}")
-    return ind
+        closest_lats = [lat_array.values[index] for index in indices]
+        closest_lons = [lon_array.values[index] for index in indices]
+        print(f"Closest indices: {indices}")
+        print(f"Given lat: {lat_point} vs {n} closest lats found: {closest_lats}")
+        print(f"Given lon: {lon_point} vs {n} closest lons found: {closest_lons}")
```
Comment (AnnikaLau) [inline iconarray/core/utilities.py:101]: This sounds a bit confusing when n=1. Should we change it to something like:
print(f"Given lat: {lat_point}. Closest {n} lat/s found: {closest_lats}")

**Review by AnnikaLau — CHANGES_REQUESTED** [review]

File: iconarray/core/utilities.py [hunk 2]
```diff
@@ -18,14 +20,14 @@ def awhere_drop(ds, cond):
 
     Parameters
     ----------
-    ds : xr.Dataset
+    ds : xarray.Dataset
```
[the remark this answers is not in the document]
Comment (leuty) [reply iconarray/core/utilities.py:24]: On top of the file `import xarray` not `import xarray as xr`. I did it consistently. Should I change it everywhere, including the import?

[the remark this answers is not in the document]
Comment (AnnikaLau) [reply iconarray/core/utilities.py:24]: Yes, I would change it everywhere in the script. I think that's standard Python style. At least we did it in other scripts, so let's be consistent.

File: iconarray/core/utilities.py (hunk 1 above)
Comment (leuty) [reply iconarray/core/utilities.py]: fixed

File: iconarray/core/utilities.py [hunk 3]
```diff
@@ -97,8 +97,8 @@ def ind_from_latlon(lon_array: xarray.DataArray, lat_array: xarray.DataArray,
         closest_lats = [lat_array.values[index] for index in indices]
         closest_lons = [lon_array.values[index] for index in indices]
         print(f"Closest indices: {indices}")
-        print(f"Given lat: {lat_point} vs {n} closest lats found: {closest_lats}")
-        print(f"Given lon: {lon_point} vs {n} closest lons found: {closest_lons}")
+        print(f"Given lat: {lat_point}. Closest {n} lat/s found: {closest_lats}")
+        print(f"Given lat: {lon_point}. Closest {n} lat/s found: {closest_lons}")
```
Comment (AnnikaLau) [inline iconarray/core/utilities.py:101]:
```suggestion
        print(f"Given lon: {lon_point}. Closest {n} lon/s found: {closest_lons}")
```

File: iconarray/core/utilities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leuty) [reply iconarray/core/utilities.py]: done

**Comment by AnnikaLau** [comment]
I created a corresponding merge request for icon-vis: https://github.com/C2SM/icon-vis/pull/39

File: iconarray/core/utilities.py [hunk 4]
```diff
@@ -35,60 +38,72 @@ def awhere_drop(ds, cond):
...
 
-    See Also
-    ----------
-    iconarray.core.utilities
+    Implementation
+    --------------
+    The function builds a KDTree from the 1D or 2D array of longitude and latitude coordinates,
+    and queries it to find the n nearest neighbors to the given point.
 
-    Examples
```
[the remark this answers is not in the document]
Comment (leuty) [reply iconarray/core/utilities.py:64]: Done

File: iconarray/core/utilities.py [hunk 5]
```diff
@@ -70,25 +90,46 @@ def ind_from_latlon(lats, lons, lat, lon, verbose=False):
     >>> lats = np.rad2deg(ds.clat.values[:])
     >>> lons = np.rad2deg(ds.clon.values[:])
```
Comment (AnnikaLau) [inline iconarray/core/utilities.py:91]:
```suggestion
    >>> lats = xr.DataArray(np.rad2deg(data.clat), coords=data.clat.coords, dims=data.clat.dims)
    >>> lons = xr.DataArray(np.rad2deg(data.clon), coords=data.clon.coords, dims=data.clon.dims)
```

Comment (AnnikaLau) [reply iconarray/core/utilities.py:91]: The old version doesn't work anymore with the changes to the script.

Comment (leuty) [reply iconarray/core/utilities.py:86]: I dont understand why exactly, but fine with me.

**Comment by leuty** [comment]
@AnnikaLau 
I realized that my approach did not work with global domains. I now use another Tree Data Structure and Haversines to compute the distances. The downside is that it adds another dependency: skikit-learn.

File: iconarray/core/utilities.py [hunk 6]
```diff
@@ -35,60 +38,97 @@ def awhere_drop(ds, cond):
     return ret
 
 
-def ind_from_latlon(lats, lons, lat, lon, verbose=False):
+def ind_from_latlon(
```
Comment (AnnikaLau) [inline iconarray/core/utilities.py:41]: If the input order is changed from lat-lon to lon-lat, the name of the function should also be adapted but I think the lat-lon order makes more sense

File: iconarray/core/utilities.py [hunk 7]
```diff
@@ -35,60 +38,97 @@ def awhere_drop(ds, cond):
...
-    verbose : bool
+    lon_array : xr.DataArray
+        A 1D or 2D xr of longitude values [radians].
+    lat_array : xr.DataArray
+        A 1D or 2D xr of latitude values [radians].
+    lon_point : float
+        The longitude value [radians] of the point to find the closest point(s) to.
+    lat_point : float
+        The latitude value [radians] of the point to find the closest point(s) to.
+    n : int, optional
```
[the remark this answers is not in the document]
Comment (leuty) [reply iconarray/core/utilities.py:64]: this should be fixed now

File: iconarray/core/utilities.py (hunk 5 above)
Comment (leuty) [reply iconarray/core/utilities.py]: This is now simplified

File: iconarray/core/utilities.py [hunk 8]
```diff
@@ -35,60 +38,134 @@ def awhere_drop(ds, cond):
...
+    Example
     ----------
     >>> # Get values of grid cell closest to coordinate
     >>> # E.g. Zürich:
     >>> lon = 8.54
     >>> lat = 47.38
-    >>> lats = np.rad2deg(ds.clat.values[:])
-    >>> lons = np.rad2deg(ds.clon.values[:])
+    >>> lats = ds.clat
+    >>> lons = ds.clon
```
Comment (AnnikaLau) [inline iconarray/core/utilities.py:86]:
```suggestion
    >>> lats = ds.clat
    >>> lons = ds.clon
    >>> if ds.clat.attrs.get('units') == 'radian':
    >>>     lats = np.rad2deg(lats)
    >>>     lons = np.rad2deg(lons)
```

[the remark this answers is not in the document]
Comment (leuty) [reply iconarray/core/utilities.py:86]: This is now done with the last commit, correct?

[the remark this answers is not in the document]
Comment (AnnikaLau) [reply iconarray/core/utilities.py:86]: yes

**Review by AnnikaLau — APPROVED** [review]
Waiting for approval of https://github.com/C2SM/icon-vis/pull/39

**Review by victoria-cherkas — APPROVED** [review]

**Pull request merged**